From: Daniel Lehman <dlehman@esri.com>
#Subject: [PATCH 2/2] ntdll: Treat INVALID_HANDLE_VALUE as success in CloseHandle.
Message-Id: <PH0PR05MB75110A531C7B889FFE1306C8DF0F9@PH0PR05MB7511.namprd05.prod.outlook.com>
Date: Mon, 14 Mar 2022 23:12:28 +0000


From 11c64857c3b5750a456f7ba54e174b586d5598b8 Mon Sep 17 00:00:00 2001
From: Daniel Lehman <dlehman@esri.com>
Date: Mon, 14 Mar 2022 12:27:40 -0700
Subject: [PATCH 2/2] ntdll: Treat INVALID_HANDLE_VALUE as success in
 CloseHandle.

https://bugs.winehq.org/show_bug.cgi?id=51529
Signed-off-by: Daniel Lehman <dlehman@esri.com>

---
visual studio remote debugging hits this exception under wine but not windows
this change would also fix the above bug
looks like only windows 10 returns success, and then only on 64-bit

---
 dlls/ntdll/tests/exception.c | 6 ------
 dlls/ntdll/unix/server.c     | 3 +++
 2 files changed, 3 insertions(+), 6 deletions(-)

diff --git a/dlls/ntdll/tests/exception.c b/dlls/ntdll/tests/exception.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/tests/exception.c
+++ b/dlls/ntdll/tests/exception.c
@@ -1247,7 +1247,6 @@ static void test_debugger(DWORD cont_status)
                 }
                 else if (stage == 17)
                 {
-                    todo_wine
                     ok(FALSE, "should not throw exception\n");
                     continuestatus = DBG_EXCEPTION_NOT_HANDLED;
                 }
@@ -3951,7 +3950,6 @@ static void test_debugger(DWORD cont_status)
                 }
                 else if (stage == 17)
                 {
-                    todo_wine
                     ok(FALSE, "should not throw exception\n");
                     continuestatus = DBG_EXCEPTION_NOT_HANDLED;
                 }
@@ -6628,7 +6626,6 @@ static void test_debugger(DWORD cont_status)
                 }
                 else if (stage == 17)
                 {
-                    todo_wine
                     ok(FALSE, "should not throw exception\n");
                     continuestatus = DBG_EXCEPTION_NOT_HANDLED;
                 }
@@ -7882,7 +7879,6 @@ static void test_debugger(DWORD cont_status)
                 }
                 else if (stage == 17)
                 {
-                    todo_wine
                     ok(FALSE, "should not throw exception\n");
                     continuestatus = DBG_EXCEPTION_NOT_HANDLED;
                 }
@@ -8698,13 +8694,11 @@ static void test_closehandle(DWORD numexc, HANDLE handle)
 
     invalid_handle_exceptions = 0;
     CloseHandle(handle);
-    todo_wine_if(handle == INVALID_HANDLE_VALUE)
     ok(invalid_handle_exceptions == numexc, "CloseHandle generated %ld exceptions, expected %ld\n",
        invalid_handle_exceptions, numexc);
 
     invalid_handle_exceptions = 0;
     pNtClose(handle);
-    todo_wine_if(handle == INVALID_HANDLE_VALUE)
     ok(invalid_handle_exceptions == numexc, "NtClose generated %ld exceptions, expected %ld\n",
        invalid_handle_exceptions, numexc);
 
diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -1071,6 +1071,9 @@ int server_get_unix_fd( HANDLE handle, unsigned int wanted_access, int *unix_fd,
 
     if (handle == INVALID_HANDLE_VALUE) return STATUS_SUCCESS;
 
+    if (handle == INVALID_HANDLE_VALUE)
+        return STATUS_SUCCESS;
+
     server_enter_uninterrupted_section( &fd_cache_mutex, &sigset );

     ret = get_cached_fd( handle, &fd, type, &access, options );
     if (ret == STATUS_INVALID_HANDLE)
-- 
2.35.1

